### PR TITLE
[SPARK-14285][SQL] Implement common type-safe aggregate functions

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

In the Dataset API, it is fairly difficult for users to perform simple aggregations in a type-safe way at the moment because there are no aggregators that have been implemented. This pull request adds a few common aggregate functions in expressions.scala.typed package, and also creates the expressions.java.typed package without implementation. The java implementation should probably come as a separate pull request. One challenge there is to resolve the type difference between Scala primitive types and Java boxed types.
## How was this patch tested?

Added unit tests for them.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+package org.apache.spark.sql.execution.aggregate
+
+import org.apache.spark.sql.expressions.Aggregator
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// This file defines internal implementations for aggregators.
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+
+class TypedSum[IN, OUT : Numeric](f: IN => OUT) extends Aggregator[IN, OUT, OUT] {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala]: this is unused at the moment

File: sql/core/src/test/java/test/org/apache/spark/sql/sources/JavaDatasetAggregatorSuite.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import org.apache.spark.sql.Encoder;
+import org.apache.spark.sql.Encoders;
+import org.apache.spark.sql.KeyValueGroupedDataset;
+import org.apache.spark.sql.expressions.Aggregator;
+import org.apache.spark.sql.test.TestSQLContext;
+
+/**
+ * Suite for testing the aggregate functionality of Datasets in Java.
+ */
+public class JavaDatasetAggregatorSuite implements Serializable {
```
Comment (rxin) [inline sql/core/src/test/java/test/org/apache/spark/sql/sources/JavaDatasetAggregatorSuite.java]: this was moved from the JavaDatasetSuite

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  val numeric = implicitly[Numeric[OUT]]
+  override def zero: OUT = numeric.zero
+  override def reduce(b: OUT, a: IN): OUT = numeric.plus(b, f(a))
+  override def merge(b1: OUT, b2: OUT): OUT = numeric.plus(b1, b2)
+  override def finish(reduction: OUT): OUT = reduction
+}
+
+
+class TypedSumDouble[IN](f: IN => Double) extends Aggregator[IN, Double, Double] {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala]: in the future we should just turn these into aggregate expressions without aggregator

**Comment by rxin** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 4]
```diff
@@ -42,7 +42,7 @@ object Window {
    * Creates a [[WindowSpec]] with the partitioning defined.
    * @since 1.4.0
    */
-  @scala.annotation.varargs
+  @_root_.scala.annotation.varargs
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: why these changes?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/java/typed.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ * :: Experimental ::
+ * Type-safe functions available for {@link Dataset} operations in Java.
+ *
+ * Scala users should use {@link org.apache.spark.sql.expressions.scala.typed}.
+ *
+ * @since 2.0.0
+ */
+@Experimental
+public class typed {
+
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/java/typed.java]: Should we add a bunch of static methods here like the `object typed`?

**Comment by cloud-fan** [comment]
Overall LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/java/typed.java (hunk 5 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/java/typed.java]: yes definitely - in a future pr though. it is not as straightforward

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 4 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: we now have a scala package in this package

**Comment by rxin** [comment]
Merging in master. Thanks. I will create tickets for follow-up work.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  override def zero: Double = 0.0
+  override def reduce(b: Double, a: IN): Double = b + f(a)
+  override def merge(b1: Double, b2: Double): Double = b1 + b2
+  override def finish(reduction: Double): Double = reduction
+}
+
+
+class TypedSumLong[IN](f: IN => Long) extends Aggregator[IN, Long, Long] {
```
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala]: Should other data types be supported (Short, Int, Float) ?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/typedaggregators.scala]: How often do you care about summing up shorts and ints but don't want to automatically coerce into longs?
